tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/TerraformUpload.java

```diff
@@ -0,0 +1,95 @@
+package com.deliveredtechnologies.maven.terraform;
+
+
+import com.deliveredtechnologies.io.CommandLine;
+import com.deliveredtechnologies.io.Executable;
+import com.deliveredtechnologies.terraform.TerraformException;
+import com.deliveredtechnologies.terraform.TerraformUtils;
+import com.deliveredtechnologies.terraform.api.TerraformOperation;
+import org.slf4j.Logger;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.Properties;
+
+
+/**
+ * upload plan files to s3 after tf:plan.
+ */
+public class TerraformUpload implements TerraformOperation<String> {
+  private Executable executable;
+
+  private Logger logger;
+
+  enum TerraformUploadParams {
+    planOutputFile,
+    sse,
+    kmsKeyId;
+  }
+
+  public TerraformUpload(String tfRootDir, Logger logger) throws IOException {
+    this(new CommandLine(tfRootDir == null  ? TerraformUtils.getDefaultTerraformRootModuleDir() : Paths.get(tfRootDir), logger), logger);
+  }
+
+  public TerraformUpload(Executable executable, Logger logger) {
+    this.executable = executable;
+    this.logger = logger;
+  }
+
+
+  /**
+   * Upload PlanFiles to s3 if -out argument starts with s3.
+   * <p>
+   *   Valid Properties:
+   *   tfRootDir - the directory containing the Terraform root module configuration; defaults to src/main/tf/{first dir found}<br>
+   *   planOutputFile - name or fully qualified s3 object name represents target location of plan files
+   *   sse - Server Side Encryption, can be either of SSE:KMS or AES:256 - defaults to AES:256
+   *   kmsKeyId - key id to encrypt s3 objects
+   * </p>
+   * @param properties  property options for Terraform plan
+   * @return            planFileName
+   * @throws TerraformException
+   */
+  @Override
+  public String execute(Properties properties) throws TerraformException, IOException {
```

Comment:
Don't change the contract to throw IOException. Trap the exception and if you can't handle it then throw TerraformException. This ensures the contract doesn't change. If you change the contract every time you have an implementation that throws a different kind of exception then it's not much of a contract. Contracts rarely change.